_posts/2017-08-25-introducing-sandboxfs.md

```diff
@@ -0,0 +1,101 @@
+
+To better illustrate this point, let's consider an example: think of a C++
+`cc_library` target that specifies `foo.cc` as a source. This target contains a
+*compilation action* that turns the `foo.cc` source file into a `foo.o` object
+file. This action will run clang, will need to read `foo.cc`, will need to read
+the transitive closure of all headers required by `foo.cc`, will need to write
+`foo.o`, and may need to write some temporary files under `/tmp/`.
+
+Under these conditions, Bazel must ensure that the action only has read-only
+access to the input files we identified and write-only access to the output and
+temporary files we expect. The reason for this is because, unless these
+conditions are enforced, Bazel cannot know that the output file really came
+from those inputs: it'd well be that the compiler went astray and read random
+files from the system, which means further builds would be inconsistent.
+
+# Current sandboxing techniques
+
+Today, Bazel uses different technologies to implement the sandboxing of
+actions. For example, on Linux, Bazel uses PID- and mount-namespaces, and on
+macOS uses symlinks and the `sandbox(7)` facility. (Note: sandboxing is
+disabled by default.)
+
+However, all these approaches have scalability and performance issues. The
+reason is that a typical build action requires hundreds, if not thousands, of
+files and directories to be mapped into the sandbox, and setting these up takes
+time. This is especially problematic on macOS: on this platform, Bazel must
+create thousands of symlinks every time it invokes an action, and this is slow.
+
+To make things worse, these approaches also have correctness issues: if
+symlinks are used, some build tools will decide to extract the real path of
+such symlinks and work off that path, which means that they may end up
+"discovering" and consuming undeclared files that are siblings of the symlink's
+target.
+
+# Enter sandboxfs
+
+To resolve these issues, we came up with the idea of implementing a FUSE file
+system that allows us to *define an arbitrary view of the host's file system
+under the mount point*. We call this **sandboxfs**.
+
+sandboxfs is efficient: the view it offers is cheaply configured at mount time
+(think that there is a single system call to configure the mount point versus
+thousands of symlink creation and deletion system calls). The view can also be
+reconfigured cheaply across different actions, avoiding the need to remount the
+FUSE file system on each action (which would also be costly).
+
+sandboxfs is correct: the view it offers is fully virtual, so sandboxfs can
+enforce arbitrary read-only and read/write access permissions on any file or
+directory it exposes. Similarly, because the view is virtual, there are no
+symlinks involved: sandboxfs exposes real files and directories to the actions,
+so those cannot reach into the original locations.
+
+# Isn't FUSE slow though?
+
+Yes, as you may know, [FUSE is slower than a real file
```

Comment:
Yikes; good catch! As a frequent comma nitpicker, I'm sad this one slipped through. In my defense, I think this sentence read different earlier and, after changing it, missed this bad comma.